contracts/HybridCustody.cdc

```diff
@@ -0,0 +1,529 @@
+    }
+
+    // Public methods anyone can call on a child account
+    pub resource interface ChildAccountPublic {
+        pub fun getParentsAddresses(): [Address]
+        pub fun getParentStatuses(): {Address: Bool}
+        pub fun getRedeemedStatus(addr: Address): Bool?
+
+        access(contract) fun setRedeemed(_ addr: Address)
+    }
+
+    // Accessible to the owner of the ChildAccount.
+    pub resource interface ChildAccountPrivate {
+        pub fun removeParent(parent: Address): Bool
+        pub fun publishToParent(
+            parentAddress: Address,
+            factory: Capability<&CapabilityFactory.Manager{CapabilityFactory.Getter}>,
+            filter: Capability<&{CapabilityFilter.Filter}>
+        )
+        pub fun giveOwnership(to: Address)
+        pub fun relinquishOwnership()
+    }
+
+    // Public methods exposed on a proxy account resource. ChildAccountPublic will share
+    // some methods here, but isn't necessarily the same
+    pub resource interface AccountPublic {
+        pub fun getPublicCapability(path: PublicPath, type: Type): Capability?
+        pub fun getAddress(): Address
+    }
+
+    // Methods only accessible to the designated parent of a ProxyAccount
+    pub resource interface AccountPrivate {
+        pub fun getCapability(path: CapabilityPath, type: Type): Capability?
+        pub fun getPublicCapability(path: PublicPath, type: Type): Capability?
+
+        access(contract) fun redeemedCallback(_ addr: Address)
+    }
+
+    // Entry point for a parent to borrow its child account and obtain capabilities or
+    // perform other actions on the child account
+    pub resource interface ManagerPrivate {
+        pub fun borrowAccount(id: UInt64): &{AccountPrivate, AccountPublic}?
+        pub fun removeChildByAddress(addr: Address)
+        pub fun removeChild(id: UInt64)
+        // TODO: Owned account methods
+    }
+
+    // Functions anyone can call on a manager to get information about an account such as
+    // What child accounts it has
+    pub resource interface ManagerPublic {
+        pub fun borrowAccountPublic(id: UInt64): &{AccountPublic}?
+        // pub fun getChildAddresses(): [Address]
+    }
+
+    /*
+    Manager
+    A resource for an account which fills the Parent role of the Child-Parent account
+    management Model. A Manager can redeem or remove child accounts, and obtain any capabilities
+    exposed by the child account to them.
+    */
+    pub resource Manager: ManagerPrivate, ManagerPublic {
+        pub let accounts: {UInt64: Capability<&{AccountPrivate, AccountPublic}>}
+        pub let addressToAccountID: {Address: UInt64}
+
+        pub let ownedAccounts: {UInt64: Capability<&{Account, ChildAccountPrivate}>}
+        pub let addressToOwnedAccountID: {Address: UInt64}
+
+        pub fun addAccount(_ cap: Capability<&{AccountPrivate, AccountPublic}>) {
+            // Is there a scenario where you are shared the same address multiple times? Seems like overkill.
```

Comment:
We'll have to make a decision on this as a community for sure. Do we want to mandate that you cannot have overlapping addresses? Or should we allow that flexibility just in case?

Reply:
Added pre checks here